src/cfn-mcp-server/awslabs/cfn_mcp_server/stack_analysis/recommendation_generator.py

```diff
@@ -0,0 +1,563 @@
+
+            chunked_resources = self._chunk_resources_by_limit(
+                filtered_resources, self.MAX_RESOURCES_PER_STACK
+            )
+
+            proposals = self._create_proposals(chunked_resources, create_templates)
+
+            return self._build_response_multi_product(
+                product_type, filtered_resources, proposals, chunked_resources, create_templates
+            )
+
+        except Exception as e:
+            logger.error(f'Error creating stack proposals: {str(e)}')
+            raise ServerError(f'Failed to create stack proposals: {str(e)}')
+
+    def _validate_augment_resources(self, resources: List[Dict]) -> None:
+        """Validate input resources for augmentation."""
+        if not resources:
+            raise ClientError('No unmanaged resources found to augment the stack.')
+
+        for resource in resources:
+            if (
+                not isinstance(resource, dict)
+                or 'ResourceType' not in resource
+                or 'ResourceIdentifier' not in resource
+            ):
+                raise ClientError(
+                    'Invalid resource format. Each resource must be a dictionary '
+                    "with 'ResourceType' and 'ResourceIdentifier' keys."
+                )
+
+    def _generate_template_name(self, prefix: str, suffix: Optional[str] = None) -> str:
+        """Generate unique template name with prefix and optional suffix."""
+        unique_id = str(uuid.uuid4())[:8]
+        if suffix:
+            return f'{prefix}-{suffix}-{unique_id}'
+        return f'{prefix}-{unique_id}'
+
+    def _wait_for_template_completion(
+        self, template_name: str, max_attempts: Optional[int] = None
+    ) -> str:
+        """Poll for template generation completion and return template body."""
+        max_attempts = max_attempts or self.MAX_TEMPLATE_POLLING_ATTEMPTS
+
+        for attempt in range(max_attempts):
+            try:
+                status_response = self.cfn_utils.describe_generated_template(template_name)
+                status = status_response.get('Status', 'UNKNOWN')
+
+                if status == self.COMPLETE_STATUS:
+                    return self.cfn_utils.get_generated_template(template_name)
+                elif status == self.FAILED_STATUS:
+                    error_reason = status_response.get('StatusReason', 'Unknown error')
+                    raise ServerError(f'Template generation failed: {error_reason}')
+                elif status in self.PENDING_STATUSES:
+                    time.sleep(self.TEMPLATE_POLLING_DELAY)
+                    continue
+                else:
+                    raise ServerError(f'Unexpected template status: {status}')
+
+            except Exception as e:
+                if attempt == max_attempts - 1:
+                    raise ServerError(
+                        f'Failed to get template after {max_attempts} attempts: {str(e)}'
+                    )
+                time.sleep(1)
```

Comment:
i think large templates, especially with lots of resources of the same type could take longer than 30 seconds. This should probably be a variable delay (Fibonacci or exponential)